### ISSUE TITLE
[RFC] MaModal component

### ISSUE THREAD

**Issue by afontcu** [holaluz/margarita#297]
- Figma link: https://www.figma.com/file/7RhweKibaUnxX6WYBl7zKW/Design-System-Beta?node-id=2566%3A0
- Release: minor

# Summary

A modal is a component with three main blocks:

* A content block, which features the "real" content of the modal. It is hidden by default, and rendered next to `body` when needed.
* A trigger block, which allows users to display the content block.
* An overlay block, which surrounds the content block. A click on this area will close the content.

It should also feature all a11y goodness described [in the spec](https://www.w3.org/TR/wai-aria-practices-1.2/#dialog_modal) (disclaimer. this docs are _informative_, not mandatory. But still).

# Basic example

```vue
<ma-modal title="Enviar la oferta por e-mail" width="small" header-type="gradient">
  <template #trigger="{openModal}">
    <button @click="openModal">Open the modal</button>
  </template>

  <!-- this will get rendered as a child of <body> -->
  <template #content="{closeModal}">
    <ma-stack>
      <p>My amazing content</p>
      <button @click="closeModal">close</button>
    </ma-stack>
  </template>
</ma-modal>
```

![image](https://user-images.githubusercontent.com/9197791/101766431-1210a500-3ae3-11eb-8080-579ba35f0475.png)

You could also use the slot to render a specific component:

```vue
<ma-modal title="My amazing modal">
  <template #trigger="{openModal}">
    <button @click="openModal">Open the modal</button>
  </template>

  <my-amazing-modal-content #content />
</ma-modal>
```

# Motivation

Modals are a common UI pattern that are hard to get right. By providing a base component from Margarita, we make sure we'll have a component that is compliant with out standards of usability, a11y, and maintainability.

# Detailed design

## Props

```js
title: { type: String, required: true }

width: { type: String, default: 'medium', validValues: ['small', 'medium', 'large'] }

headerType: { type: String, default: 'white', validValues: ['white', 'gradient'] }
// to render a gradient heading
```

## Slots

```
#trigger (openModal: Function)
#content (closeModal: Function)
```

## Emits

```
'open'
'close'
```

The `close` emit will be triggered when:
1) the `closeModal` function is executed
2) user clicks on the overlay
3) user clicks escape

## DOM output

The component will take everything from the `#content` slot and move it next to the end tag of `body`. This way, we avoid potential issues when it comes to absolute/relative positioning, and z-indexes. The implementation of this feature can be done by hand, or using [portal-vue](https://github.com/LinusBorg/portal-vue).

If MaModal contains a component, it will remain a logical child component of the MaModal's parent. 

~This also means that injections from a parent component work as expected, and that the child component will be nested below the parent component in the Vue Devtools, instead of being placed where the actual content moved to.~
update: I don't think this is doable in Vue 2 without a lot of magic. This means that if the slot content of MaModal contains components, they will show up as children of MaModal in the devtools.

(Simplified) resulting DOM structure from the previous example:

```html
<html>
  <body>
    <button>Open the modal</button>

    <!-- modal structure is totally made up. Simply notice where it gets rendered -->
    <div class="ma-modal">
      <p class="ma-modal__heading">Enviar la oferta por e-mail</p>
      <div class="ma-stack">
        <p>My amazing content</p>
        <button>close</button>
      </div>
    </div>
  </body>
</html>
```

## Styling
MaModal aims to provide little-to-no styling. It will render the overlay, center the modal, pad its inner content and manage its size, but nothing beyond this point. MaModal does not aim to lay out its inner content other than heading, so consumer should be responsible for it.

In terms of width, `width` prop will set a max width for the modal, and shrink if needed.

In terms of height, the component will grow to wrap its content, and overflow (with scroll) if content grows higher than available vertical space.

MaModal heading will render the provided title value and also a close button which will close the modal. The heading will remain visible even when vertical overflow is present.

## Accessibility

As mentioned, it should feature all a11y goodness described [in the spec](https://www.w3.org/TR/wai-aria-practices-1.2/#dialog_modal). In particular, autofocus on open, focus trap, and appropriate ARIA roles must be there.

# Adoption strategy

While this RFC does not introduce a breaking change, adopting MaModal in projects might require several modifications (depending on current usage). That being said, I cannot foresee any current scenario where MaModal isn't able to match current features of our modals.

# Unresolved questions

- ~Should we allow for multiple MaModal elements to be opened at the same time?~ update: No
- ~Should the `size` prop accept a ResponsiveProp value such as `['small', 'large']`?~ update: No
- ~Are there situations where we need to know if the modal is open, for instance in the `#trigger` slot? Should this info be provided to the slot?~ update: Not for now.

**Comment by icurdu** [comment]
Such a good job!

I have 2 questions:
1. Should we specify in the summary, that the modal content has 2 indivisible parts? I'm talking about the modal-header and the modal-body. The modal header always has the same structure and way of behaving (title + close button, and sticks to the top while overflow-scroll).
2. I'm not sure about setting the gradient as a prop. We think about the gradient as a "type of header", and actually we have two types of header (white-default and gradient) but we don't know in the future if we will need more styles.

**Comment by afontcu** [comment]
thanks for your feedback! 🙌 

> Should we specify in the summary, that the modal content has 2 indivisible parts? I'm talking about the modal-header and the modal-body. The modal header always has the same structure and way of behaving (title + close button, and sticks to the top while overflow-scroll).

Sure! That was the intention, maybe it's not clear enough. I'll try to rephrase it.


> I'm not sure about setting the gradient as a prop. We think about the gradient as a "type of header", and actually we have two types of header (white-default and gradient) but we don't know in the future if we will need more styles.

You mean using something like

```vue
<ma-modal header-type="gradient" />
```

instead? I guess it's ok, too! I like boolean (true/false) prop where possible because they convey that we only have two possibilities, but if you know that we might need three in a near future, we can certainly define it as a "regular" attribute :)

**Comment by alexcastelo95** [comment]
It looks good.

+1 to add some flexibility to the `header-type` in the near future. I understand that if we don't have more than 2 possibilities now we can manage it with a boolean, but I'm sure that at some point someone will ask us for another type of header.

**Comment by edufarre** [comment]
> Should we allow for multiple MaModal elements to be opened at the same time?

I cannot think of a real-world, case but in my opinion, if we need to display/have more than one modal opened at the same time, it is kinda weird, isn't it? Would that mean that the way we are currently displaying whatever info we want to display is not the proper/best tone?

> Should the size prop accept a ResponsiveProp value such as ['small', 'large']?

In this case, I would say that the `width` property would be enough. Just because if we allow to have a responsive prop as you suggested, that could lead to having inconsistency between the modals used on the targeted application. IMHO, if a modal has `width="medium"` should be `medium` in desktop AND mobile.

> Are there situations where we need to know if the modal is open, for instance in the #trigger slot? Should this info be provided to the slot?

AKAIK, if we display a modal, we do not allow the user to interact with the other page sections/components. However, I can not come up with a scenario where we could need this. 
Shall we let this question for a second iteration?

> title: { type: String, required: true }
Not all our modals have 'title', I believe the title property should not be required.

In a nutshell, the RFC looks great, let's discuss these few points and go for it! 🚀

**Comment by afontcu** [comment]
> Would that mean that the way we are currently displaying whatever info we want to display is not the proper/best tone?

Yeah, agree! Let's not explicitly allow this.

> AKAIK, if we display a modal, we do not allow the user to interact with the other page sections/components. However, I can not come up with a scenario where we could need this.

Yeah, that make sense. Having the `on` emit seems enough.


> Not all our modals have 'title', I believe the title property should not be required.

Oh, interesting. Can you confirm this @icurdu? I assumed all modals will have a title, but now I think I might have make this up 🤔 


thanks for your thoughts!

**Comment by afontcu** [comment]
> Oh, interesting. Can you confirm this @icurdu? I assumed all modals will have a title, but now I think I might have make this up 🤔

Confirmed: they will all have a title, so we can keep it as a required prop 🎉
